libraries/chain/controller.cpp

```diff
@@ -423,7 +423,8 @@ struct controller_impl {
          EOS_ASSERT( root_id == log_head_id, fork_database_exception, "fork database root does not match block log head" );
       } else {
          EOS_ASSERT( fork_db.root()->block_num == lib_num, fork_database_exception,
```

Comment:
Do we hit this `else` block *only* when we use the `empty_block_log`?

Reply:
Yes. Otherwise blog would have a head.

Reply:
Hum, so if the block log doesn't have a head, it means it doesn't have any block, right. So it doesn't have a first block either. I think logically `blog.first_block_num()` should return an empty `std::optional<uint32_t>`. And because we don't have a `lib`, `lib_num` should also be an empty `std::optional`. Here it seems like we are jumping through hoops to create a value matching what `fork_db.root()->block_num` will be, right? Or am I still missing something?

Reply:
I think you make an excellent point. We could certainly look into making a better invariant for block log around first block in a separate PR to main. However, as the software currently stands, it supports the idea of a first block even when there is nothing in the block log. That is true of basic block log and pruned block log as well. Changing that here would very likely break a number of things.